File: date/3.14/3.cpp
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        std::unordered_map<char, int> mp;
        int begin = 0, maxL = 0;
        for (int i = 0; i < s.length(); i++)
        {
            if (mp.find(s[i]) != mp.end())
            {
                if (mp[s[i]] + 1 < begin)
                {
                    begin = begin;
                }
                else
                {
                    begin = mp[s[i]] + 1;
                }
                mp.erase(s[i]);
            }
            mp[s[i]] = i;
            int l = i - begin + 1;
            if (l > maxL)
            {
                maxL = l;
            }
        }
        return maxL;
    }
};

File: date/3.14/3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        hs={}
        begin=0
        maxL=0
        for i in range(len(s)):
            if s[i] in hs:
                if hs[s[i]]+1<begin:
                    begin=begin
                else:
                    begin=hs[s[i]]+1
                del hs[s[i]]
            hs[s[i]]=i
            l=i-begin+1
            if l>maxL:
                maxL=l
        return maxL